util/messaging.py

```diff
@@ -0,0 +1,234 @@
+        'coord' - A thread coordinator to automatically stop the threads
+        '''
+        threads = []
+        for routine in [self._send_loop, self._receive_loop]:
+            thread = Thread(target=routine, args=(session,coord))
+            thread.daemon = True
+            thread.start()
+            threads.append(thread)
+        log.debug('Started queue threads')
+        return threads
+
+    def close_queues(self, session):
+        '''
+        Closes all queues of this instance.
+        'session' - The current session
+        '''
+        session.run(self._inbound_close)
+        for close in self._outbound_closes:
+            session.run(close)
+        log.debug('Closed queue threads')
+
+    def call_async(self, job, task, function, callback, *args):
+        '''
+        Performs an asynchronous (non-blocking) RPC call to a remote node's function.
+        If 'job' and 'task' address the local node, the call is performed locally.
+        'job' - Job name of the local cluster node
+        'task' - Task index of the local cluster node
+        'function' - Name of the remote function (member of class derived from ClusterMessagingClient)
+        'callback' - Will be called with return value tuple of remote function as arguments,
+            if RPC call succeeded. Omitting it (None) results in dependency free state
+            (the local client can safely be stopped/ended).
+            Can also be an Event object (used internally by 'call').
+        'args' - Arguments to the remote function.
+        Returns a callback id (0 in case of no callback).
+        '''
+        # target client id (shared_name for the apropos queue)
+        id = self._id_pattern % (job, task)
+        if self.id == id:
+            # local call - no queueing needed
+            log.step('Local call of function: %s, arguments: %r' % (function, args))
+            results = self._call(function, args)
+            if callback:
+                callback(results)
+            return -1
+        log.step('Calling worker: %s, function: %s, arguments: %r' % (id, function, args))
+        # default value in case of no callback, if not waiting for responseno
+        # and no response should be produced on the receiving end.
+        response_id = 0
+        if callback:
+            with self._response_id_lock:
+                # acquire a new response id
+                response_id = self._response_id_counter = self._response_id_counter + 1
+                # register callback or Event object as 'waiting for response'
+                self._waiting_for_response[response_id] = callback
+        # en-pickle header and call arguments
+        content = pickle.dumps((self.id, False, response_id, function, args))
+        # get enqueue op of outbound queue of the targeted client
+        enqueue = self._outbound_enqueues[id]
+        # enqueue locally for sending
+        self._to_send.put((enqueue, { self._ph_content: content }))
```

Comment:
All the callsites like `self._to_send.put()` does the same so far, maybe there would be some value in having something that enforces the way we do that call, to make sure further evolution do not slip through?

Reply:
Thought the same once - properly normalizing it costs more lines of code. :-)